Auth: Handle dangling permissions

markylaing opened the pull request:
The `permissions` table has `entity_id`, `entity_type`, and `entitlement` columns. When a permission is created, we validate that the entity exists and get its ID. However, if the entity is deleted the permission is left dangling.

Dangling permissions could never have been displayed or used because we can't construct a URL for them for use in the API responses. However, not handling them led to some strange behaviour in the OpenFGA datastore (see https://github.com/canonical/lxd/pull/12976#issuecomment-1969918212) and of course we don't want the permissions table to grow indefinitely.

This PR adds logic to lazily delete dangling permissions as they are encountered.

tomponline commented on lxd/db/cluster/permissions.go:
```diff
@@ -109,6 +109,41 @@ func GetPermissionEntityURLs(ctx context.Context, tx *sql.Tx, permissions []Perm
+	}
+
+	var q string
+	if len(args) == 1 {
```
This is overly-complex in my view, why not always use the latter form with just a single item?

markylaing replied:
Looking back at this I initially misread the `query.Params` function and thought it didn't handle the single argument case correctly. I'll change it.

markylaing replied:
Done.

tomponline commented on lxd/db/cluster/permissions.go:
```diff
@@ -109,6 +109,41 @@ func GetPermissionEntityURLs(ctx context.Context, tx *sql.Tx, permissions []Perm
+	}
+
+	if len(args) != int(rowsAffected) {
+		return fmt.Errorf("Failed to delete expected number of dangling permissions on clean up (expected `%d`, got `%d`)", len(args), rowsAffected)
```
No need to quite %d

markylaing replied:
Done.

tomponline reviewed (commented):
My main question with this PR is can we remove the permissions when the entity is deleted? Its great we are detecting any dangling ones and handling them to avoid access control issues - but leaving orphaned entries in the DB by design seems rather unfortunate - and something we would have considered to be a bug previously.

tomponline commented:
Also, are there any risks of a dangling permission being re-used for a new entity of the same type and name & project?

markylaing commented:
> Also, are there any risks of a dangling permission being re-used for a new entity of the same type and name & project?

No, when the new entity is added its ID will be different. The embedded driver wont be able to get a URL for the permission with the ID of the deleted entity because it will not be able to query for it's name/project etc. by ID. Also, when adding a permission to a group the URL is parsed to get an entity ID, as the old entity has been deleted, the new permission can only reference the new entity.

> My main question with this PR is can we remove the permissions when the entity is deleted? Its great we are detecting any dangling ones and handling them to avoid access control issues - but leaving orphaned entries in the DB by design seems rather unfortunate - and something we would have considered to be a bug previously.

There are a few approaches here and non are perfect:
1. Use database triggers to delete permissions when any entity is deleted, or when a permission is no longer associated with any groups. This would require an awful lot of triggers but would be handled by the database automatically leading to less mental overhead (e.g. not having to remember to delete permissions all over the code base). However, we should still validate all of the permissions before using them for authorization or as an API response because it's possible that we missed a trigger.
2. Lazily clean up permissions as they are requested and ignore dangling permissions during authorization. This requires validating each permission during authorization but I think we should be doing this anyway in case we missed one (also, we currently need to get the URL of the entity to generate the OpenFGA tuple so validation is more or less a consequence of that).
3. Use a background task to delete dangling permissions. This approach in isolation is not enough since a permission can become orphaned in-between refresh intervals.
4. Actively delete permissions manually when we delete an entity. This approach was necessary when we implemented the remote OpenFGA authorizer as we need to create/rename/delete tuples in the remote store. The calls are still present in the code base and I am planning to remove them (#12975). My feeling is this approach negates the benefit of building the embedded driver, as there is now another mental overhead when deleting an entity (entity specific clean up - fine, delete from DB -fine, remember to emit a lifecycle event - error prone, remember to delete any associated permissions - error prone). This has already led to some issues (see https://github.com/lxc/incus/commit/a7431c6973da17a25b190d1038b71b5aa920c4a9, in this case not all storage volume types were being added/removed).

I realised this morning that there is another case where permissions can be dangling. If a permission is added to a group, the permission is created and an association is made between the group and the permission. If the group is subsequently deleted, the association between the group and permission is also deleted but the permission itself is not deleted.

Overall, my feeling is that the best approach will be a combination of 2 and 3. 2 ensures that dangling permissions are never used for authorization, nor are they ever shown via the API. 3 can additionally delete permissions that don't belong to any groups and will ensure clean up happens regularly enough that the permissions table doesn't grow too much. I kind of think of this like a garbage collector that optimistically cleans up when something goes out of scope, but also does a stop-the-world periodically.

markylaing commented:
@tomponline I've implemented the changes we discussed:
* The `permissions` table is removed and `auth_groups_permissions` contains permission data directly. Permissions associated with a group will now automatically be deleted via foreign key relationship.
* On daemon start up, if we are the leader we apply triggers to the database that automatically clean up dangling permissions and warnings.
* When getting permissions URLs we still check in case any are dangling and ignore them.
* A cluster task has been added to check for dangling permissions and delete them. It logs a warning if any are found (because these should have been deleted by triggers). If if fails to delete the the dangling permissions it creates a warning for each permission so that they can be manually deleted.

markylaing commented:
@tomponline I've implemented the changes we discussed:
* The `permissions` table is removed and `auth_groups_permissions` contains permission data directly. Permissions associated with a group will now automatically be deleted via foreign key relationship.
* On daemon start up, if we are the leader we apply triggers to the database that automatically clean up dangling permissions and warnings.
* When getting permissions URLs we still check in case any are dangling and ignore them.
* A cluster task has been added to check for dangling permissions and delete them. It logs a warning if any are found (because these should have been deleted by triggers). If if fails to delete the the dangling permissions it creates a warning for each permission so that they can be manually deleted.

markylaing commented:
I've just now remembered that we can get rid of a lot of warning deletion logic so I'll keep as draft until I've done that.

tomponline commented:
> * A cluster task has been added to check for dangling permissions and delete them. It logs a warning if any are found (because these should have been deleted by triggers). If if fails to delete the the dangling permissions it creates a warning for each permission so that they can be manually deleted.

Do we need this really?

markylaing commented:
> > * A cluster task has been added to check for dangling permissions and delete them. It logs a warning if any are found (because these should have been deleted by triggers). If if fails to delete the the dangling permissions it creates a warning for each permission so that they can be manually deleted.
> 
> Do we need this really?

We can go back to cleaning up opportunistically if you like. My thinking was to remove any logic from the API endpoints that aren't specifically required. I suppose I can write a clean up function and defer call to it in the API handlers?

tomponline commented:
> We can go back to cleaning up opportunistically if you like.

That's what the triggers were to prevent right?

markylaing commented:
> > We can go back to cleaning up opportunistically if you like.
> 
> That's what the triggers were to prevent right?

Yes. Basically I added the task so that LXD complains fairly loudly if the triggers aren't working as expected.

tomponline commented:
> Yes. Basically I added the task so that LXD complains fairly loudly if the triggers aren't working as expected.

I don't think we need a task if the triggers are there. We could log a warning if the access checker detects them when its ignoring them perhaps?

markylaing commented:
> > Yes. Basically I added the task so that LXD complains fairly loudly if the triggers aren't working as expected.
> 
> I don't think we need a task if the triggers are there. We could log a warning if the access checker detects them when its ignoring them perhaps?

Sure happy to do that :)

markylaing commented:
@tomponline this PR is causing some of the clustering tests to fail. Sometimes when setting up another cluster member it fails to start up. I think it's to do with the logic for when to apply the triggers in this commit: https://github.com/canonical/lxd/pull/12992/commits/6b2a6119e252d1158dfc9790722385cbafb2d8d6

On startup I'm calling `gateway.LeaderAddress` and checking if the value is the same as the configured cluster address. If so I'm applying the triggers. Perhaps we can take a look together in our 1-2-1 on Monday. Have a good weekend :)

markylaing commented:
@tomponline I've moved the call to apply the triggers into the cluster DB `Open` function. It seems to work locally so we'll see if CI complains.

Regarding the triggers themselves, when we do #12928 I would like to add methods to the entity interface `OnDeleteTriggerName` and `OnDeleteTriggerSQL` so that when applying triggers we can check it exists with `SELECT sql FROM sqlite_master WHERE type = 'trigger' AND name = '<trigger_name>'` and only apply it if it doesn't exist or if the SQL has changed. 

Regarding warnings. It looks like the warning for the proxy device is being deleted. As far as I can tell, that's the only place in the codebase where any warnings are deleted except for a task which prunes warnings that have been resolved, and via the API directly with a UUID. Since we don't have an entity type for an instance device yet, I didn't want to add that here for the sole purpose of tidying up the warnings system.

tomponline commented:
> @tomponline I've moved the call to apply the triggers into the cluster DB `Open` function. It seems to work locally so we'll see if CI complains.
> 
> Regarding the triggers themselves, when we do #12928 I would like to add methods to the entity interface `OnDeleteTriggerName` and `OnDeleteTriggerSQL` so that when applying triggers we can check it exists with `SELECT sql FROM sqlite_master WHERE type = 'trigger' AND name = '<trigger_name>'` and only apply it if it doesn't exist or if the SQL has changed.
> 
> Regarding warnings. It looks like the warning for the proxy device is being deleted. As far as I can tell, that's the only place in the codebase where any warnings are deleted except for a task which prunes warnings that have been resolved, and via the API directly with a UUID. Since we don't have an entity type for an instance device yet, I didn't want to add that here for the sole purpose of tidying up the warnings system.

Sounds all good thanks

tomponline commented on lxd/db/cluster/permissions.go:
```diff
@@ -92,3 +94,32 @@ func GetPermissionEntityURLs(ctx context.Context, tx *sql.Tx, permissions []Perm
 }
+
+// DeletePermissions deletes the given permissions.
+func DeletePermissions(ctx context.Context, tx *sql.Tx, danglingPermissions []Permission) error {
```
Do we need this function anymore?

markylaing replied:
No we don't. The linter should have complained about this as it's unused :thinking:

tomponline commented on test/suites/auth.sh:
```diff
@@ -49,7 +49,11 @@ test_authorization() {
+
+  # Test permission is removed automatically when instance is removed.
+  lxc auth group permission add test-group instance c1 can_exec project=default # Valid
   lxc rm c1
```
Please always add the `--force` flag to the test suite to ensure if it is running it wont hang the suite.

tomponline commented on lxd/db/cluster/open_test.go:
```diff
@@ -65,7 +65,7 @@ func TestEnsureSchema_ClusterNotUpgradable(t *testing.T) {
 			},
 			false,
-			"this node's version is behind, please upgrade",
+			"Failed to ensure schema: this node's version is behind, please upgrade",
```
As we're changing this, please can we also say "member" instead of "node"?

tomponline commented on lxd/db/cluster/open.go:
```diff
@@ -63,7 +63,7 @@ func EnsureSchema(db *sql.DB, address string, dir string) (bool, error) {
 
 	backupDone := false
-	hook := (func(ctx context.Context, version int, tx *sql.Tx) error {
+	hook := func(ctx context.Context, version int, tx *sql.Tx) error {
```
What's the change in brackets here?

tomponline commented on lxd/db/cluster/open.go:
```diff
@@ -155,9 +155,20 @@ func EnsureSchema(db *sql.DB, address string, dir string) (bool, error) {
 
 	var initial int
-	err := query.Retry(context.TODO(), func(_ context.Context) error {
+	err := query.Retry(context.TODO(), func(ctx context.Context) error {
```
The commit message should be about triggers not warnings right?

tomponline reviewed (commented):
There doesn't appear to be any error wrapping in that last commit despite what the commit message says

markylaing commented:
> There doesn't appear to be any error wrapping in that last commit despite what the commit message says

I've reworded the commits to separate it out properly. This was an oversight while rebasing.

markylaing commented on lxd/util/version.go:
```diff
@@ -31,5 +31,5 @@ func CompareVersions(version1, version2 [2]int) (int, error) {
 	}
 
-	return -1, fmt.Errorf("nodes have inconsistent versions")
+	return -1, fmt.Errorf("cluster members have inconsistent versions")
```
I'll capitalise this.
